File: patterns/Behavioral/Command/JavaScript/Command.js
// Command interface declares a method for executing a command.
class Command {
  execute() {}
}

// Some commands can implement simple operations on their own.
class SimpleCommand extends Command {
  constructor(payload) {
    super();
    this.payload = payload;
  }

  execute() {
    console.log(`SimpleCommand: See, I can do simple things like printing (${this.payload})`);
  }
}

// However, some commands can delegate more complex operations to other objects, called "receivers."
class ComplexCommand extends Command {
  constructor(receiver, a, b) {
    super();
    this.receiver = receiver;
    this.a = a;
    this.b = b;
  }

  execute() {
    console.log("ComplexCommand: Complex stuff should be done by a receiver object.");
    this.receiver.doSomething(this.a);
    this.receiver.doSomethingElse(this.b);
  }
}

// The Receiver classes contain some important business logic. They know how to perform all kinds of operations, associated with carrying out a request. In fact, any class may serve as a Receiver.
class Receiver {
  doSomething(a) {
    console.log(`Receiver: Working on (${a}).`);
  }

  doSomethingElse(b) {
    console.log(`Receiver: Also working on (${b}).`);
  }
}

// The Invoker is associated with one or several commands. It sends a request to the command.
class Invoker {
  setOnStart(command) {
    this.onStart = command;
  }

  setOnFinish(command) {
    this.onFinish = command;
  }

  doSomethingImportant() {
    console.log("Invoker: Does anybody want something done before I begin?");
    if (this.onStart instanceof Command) {
      this.onStart.execute();
    }

    console.log("Invoker: ...doing something really important...");

    console.log("Invoker: Does anybody want something done after I finish?");
    if (this.onFinish instanceof Command) {
      this.onFinish.execute();
    }
  }
}

// The client code can parameterize an invoker with any commands.
const invoker = new Invoker();
invoker.setOnStart(new SimpleCommand("Say Hi!"));
const receiver = new Receiver();
invoker.setOnFinish(new ComplexCommand(receiver, "Send email", "Save report"));

invoker.doSomethingImportant();

File: patterns/Behavioral/Command/Go/Command.go
package main

import "fmt"

/*
|--------------------------------------------------------------------------
| Command Design Pattern
|--------------------------------------------------------------------------
| This example demonstrates the Command Design Pattern
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Behavioral/Command
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/milendenev0912/design-patterns
|--------------------------------------------------------------------------
*/

// Command interface declares a method for executing a command.
type Command interface {
	Execute()
}

// SimpleCommand implements the Command interface for simple operations.
type SimpleCommand struct {
	payload string
}

func NewSimpleCommand(payload string) *SimpleCommand {
	return &SimpleCommand{payload: payload}
}

func (s *SimpleCommand) Execute() {
	fmt.Printf("SimpleCommand: See, I can do simple things like printing (%s)\n", s.payload)
}

// ComplexCommand delegates more complex operations to a Receiver.
type ComplexCommand struct {
	receiver *Receiver
	a        string
	b        string
}

func NewComplexCommand(receiver *Receiver, a string, b string) *ComplexCommand {
	return &ComplexCommand{receiver: receiver, a: a, b: b}
}

func (c *ComplexCommand) Execute() {
	fmt.Println("ComplexCommand: Complex stuff should be done by a receiver object.")
	c.receiver.DoSomething(c.a)
	c.receiver.DoSomethingElse(c.b)
}

// Receiver contains business logic and knows how to perform operations.
type Receiver struct{}

func (r *Receiver) DoSomething(a string) {
	fmt.Printf("Receiver: Working on (%s.)\n", a)
}

func (r *Receiver) DoSomethingElse(b string) {
	fmt.Printf("Receiver: Also working on (%s.)\n", b)
}

// Invoker is associated with commands and sends requests to them.
type Invoker struct {
	onStart  Command
	onFinish Command
}

func (i *Invoker) SetOnStart(command Command) {
	i.onStart = command
}

func (i *Invoker) SetOnFinish(command Command) {
	i.onFinish = command
}

func (i *Invoker) DoSomethingImportant() {
	fmt.Println("Invoker: Does anybody want something done before I begin?")
	if i.onStart != nil {
		i.onStart.Execute()
	}

	fmt.Println("Invoker: ...doing something really important...")

	fmt.Println("Invoker: Does anybody want something done after I finish?")
	if i.onFinish != nil {
		i.onFinish.Execute()
	}
}

// Client code parameterizes an invoker with commands.
func main() {
	invoker := &Invoker{}
	invoker.SetOnStart(NewSimpleCommand("Say Hi!"))
	receiver := &Receiver{}
	invoker.SetOnFinish(NewComplexCommand(receiver, "Send email", "Save report"))

	invoker.DoSomethingImportant()
}